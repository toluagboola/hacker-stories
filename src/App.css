body {
  font-family: "IBM Plex Mono", -apple-system;
}

#root {
  max-width: 700px;
  background: #161b22;
}

.title,
strong {
  color: #fff;
}

.input {
  color: #fff;
  background: inherit;
}

a:hover {
  color: #3273dc;
  text-decoration: underline;
}

.dismiss {
  color: #fff;
  background: transparent;
}

.dismiss:hover {
  color: #fff;
}

.button[disabled] {
  background: inherit;
}
