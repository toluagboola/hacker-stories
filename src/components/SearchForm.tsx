import React from "react";
import { Search } from "react-feather";

type SearchFormProps = {
	searchTerm: string;
	onSearchInput: (event: React.ChangeEvent<HTMLInputElement>) => void;
	onSearchSubmit: (event: React.FormEvent<HTMLFormElement>) => void;
};

const SearchForm = ({
	searchTerm,
	onSearchInput,
	onSearchSubmit,
}: SearchFormProps) => (
	<form onSubmit={onSearchSubmit}>
		<InputWithLabel
			id="search"
			value={searchTerm}
			isFocused
			onInputChange={onSearchInput}
		>
			Search
		</InputWithLabel>
		<button
			className="button mt-2 is-primary search-story"
			type="submit"
			disabled={!searchTerm}
		>
			Search
		</button>
	</form>
);

type InputWithLabelProps = {
	id: string;
	value: string;
	type?: string;
	onInputChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
	isFocused?: boolean;
	children?: React.ReactNode;
};

const InputWithLabel = ({
	id,
	value,
	type = "text",
	onInputChange,
	isFocused,
	children,
}: InputWithLabelProps) => {
	const inputRef = React.useRef<HTMLInputElement>(null!);

	React.useEffect(() => {
		if (isFocused && inputRef.current) {
			inputRef.current.focus();
		}
	}, [isFocused]);

	return (
		<div className="control has-icons-left">
			<span className="icon is-left px-2">
				<Search />
			</span>
			<input
				ref={inputRef}
				className="input mb-4"
				type={type}
				id={id}
				value={value}
				onChange={onInputChange}
				autoFocus={isFocused}
				placeholder="Search for a story"
			/>
		</div>
	);
};

export default SearchForm;
