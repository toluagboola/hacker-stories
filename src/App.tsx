import React, {
  useState,
  useEffect,
  useRef,
  useReducer,
  useCallback,
} from "react";
import axios from "axios";
import "bulma/css/bulma.min.css";
import "./App.css";
import { Check } from "react-feather";

type Story = {
  objectID: string;
  url: string;
  title: string;
  author: string;
  num_comments: number;
  points: number;
};

type Stories = Array<Story>;

type StoriesState = {
  data: Stories;
  isLoading: boolean;
  isError: boolean;
};

interface StoriesFetchInitAction {
  type: "STORIES_FETCH_INIT";
}

interface StoriesFetchSuccessAction {
  type: "STORIES_FETCH_SUCCESS";
  payload: Stories;
}

interface StoriesFetchFailureAction {
  type: "STORIES_FETCH_FAILURE";
}

interface StoriesRemoveAction {
  type: "REMOVE_STORY";
  payload: Story;
}

type StoriesAction =
  | StoriesFetchInitAction
  | StoriesFetchSuccessAction
  | StoriesFetchFailureAction
  | StoriesRemoveAction;

// Handle state transitions
const storiesReducer = (state: StoriesState, action: StoriesAction) => {
  switch (action.type) {
    case "STORIES_FETCH_INIT":
      return {
        ...state,
        isLoading: true,
        isError: false,
      };
    case "STORIES_FETCH_SUCCESS":
      return {
        ...state,
        isLoading: false,
        isError: false,
        data: action.payload,
      };
    case "STORIES_FETCH_FAILURE":
      return {
        ...state,
        isLoading: false,
        isError: true,
      };
    case "REMOVE_STORY":
      return {
        ...state,
        data: state.data.filter(
          (story) => action.payload.objectID !== story.objectID
        ),
      };
    default:
      throw new Error();
  }
};

// Persist latest search query
const useSemiPersistentState = (
  key: string,
  initialState: string
): [string, (newValue: string) => void] => {
  const [value, setValue] = useState(localStorage.getItem(key) || initialState);
  const isMounted = useRef(false);

  useEffect(() => {
    if (isMounted.current) {
      isMounted.current = true;
    } else {
      localStorage.setItem(key, value);
    }
  }, [value, key]);

  return [value, setValue];
};

// const getSumComments = (stories) => {
//   return stories.data.reduce((result, value) => result + value.num_comments, 0);
// };

const API_ENDPOINT = "https://hn.algolia.com/api/v1/search?query=";

// App component
const App = () => {
  const [searchTerm, setSearchTerm] = useSemiPersistentState("search", "React");
  const [url, setUrl] = useState(`${API_ENDPOINT}${searchTerm}`);
  const [stories, dispatchStories] = useReducer(storiesReducer, {
    data: [],
    isLoading: false,
    isError: false,
  });

  // Handle HTTP requests
  const handleFetchStories = useCallback(async () => {
    dispatchStories({ type: "STORIES_FETCH_INIT" });

    try {
      const result = await axios.get(url);
      dispatchStories({
        type: "STORIES_FETCH_SUCCESS",
        payload: result.data.hits,
      });
    } catch {
      dispatchStories({ type: "STORIES_FETCH_FAILURE" });
    }
  }, [url]);

  // Handle side effects
  useEffect(() => {
    handleFetchStories();
  }, [handleFetchStories]);

  // Handle story removal from list
  const handleRemoveStory = useCallback((item: Story) => {
    dispatchStories({
      type: "REMOVE_STORY",
      payload: item,
    });
  }, []);

  const handleSearchInput = (event: React.ChangeEvent<HTMLInputElement>) => {
    setSearchTerm(event.target.value);
  };

  // Handle form submission
  const handleSearchSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    setUrl(`${API_ENDPOINT}${searchTerm}`);
    event.preventDefault();
  };

  // const sumComments = getSumComments(stories);

  return (
    <div className="container has-text-centered has-text-white">
      <h1 className="title is-3">Hacker Stories</h1>
      {/*<h3 className="subtitle has-text-link">{sumComments} total comments</h3>*/}

      <SearchForm
        searchTerm={searchTerm}
        onSearchInput={handleSearchInput}
        onSearchSubmit={handleSearchSubmit}
      />

      <hr />

      {stories.isError && <p>Something went wrong...</p>}

      {stories.isLoading ? (
        <p>Loading...</p>
      ) : (
        <List list={stories.data} onRemoveItem={handleRemoveStory} />
      )}
    </div>
  );
};

type InputWithLabelProps = {
  id: string;
  value: string;
  type?: string;
  onInputChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  isFocused?: boolean;
  children?: React.ReactNode;
};

const InputWithLabel = ({
  id,
  value,
  type = "text",
  onInputChange,
  isFocused,
  children,
}: InputWithLabelProps) => {
  const inputRef = React.useRef<HTMLInputElement>(null!);

  React.useEffect(() => {
    if (isFocused && inputRef.current) {
      inputRef.current.focus();
    }
  }, [isFocused]);

  return (
    <div className="control">
      <input
        ref={inputRef}
        className="input mb-4"
        type={type}
        id={id}
        value={value}
        onChange={onInputChange}
        autoFocus={isFocused}
        placeholder="Search for a story"
      />
    </div>
  );
};

// class InputWithLabel extends React.Component {
//   constructor(props) {
//     super(props);
//     this.inputRef = React.createRef();
//   }

//   componentDidMount() {
//     if (this.props.isFocused) {
//       this.inputRef.current.focus();
//     }
//   }

//   render() {
//     const { id, type = "text", value, isFocused, onInputChange } = this.props;
//     return (
//       <div className="control">
//         <input
//           ref={this.inputRef}
//           className="input mb-4"
//           type={type}
//           id={id}
//           value={value}
//           onChange={onInputChange}
//           autoFocus={isFocused}
//           placeholder="Search for a story"
//         />
//       </div>
//     );
//   }
// }

type ListProps = {
  list: Stories;
  onRemoveItem: (item: Story) => void;
};

const List = ({ list, onRemoveItem }: ListProps) => (
  <div className="py-2">
    {list.map((item) => (
      <Item key={item.objectID} item={item} onRemoveItem={onRemoveItem} />
    ))}
  </div>
);

type ItemProps = {
  item: Story;
  onRemoveItem: (item: Story) => void;
};

const Item = ({ item, onRemoveItem }: ItemProps) => {
  return (
    <div className="py-3 has-text-left">
      <p>
        <a className="is-size-5 has-text-weight-medium" href={item.url}>
          {item.title}
        </a>
      </p>
      <p className="pb-1 is-capitalized">
        <strong>Author:</strong> {item.author}
      </p>
      <p className="pb-1 is-capitalized">
        <strong>Comments:</strong> {item.num_comments}
      </p>
      <p className="pb-1 is-capitalized">
        <strong>Points:</strong> {item.points}
      </p>
      <button
        type="button"
        className="button is-small mt-2 dismiss"
        onClick={() => onRemoveItem(item)}
      >
        <span className="icon">
          <Check />
        </span>
        <span>Check</span>
      </button>
    </div>
  );
};

type SearchFormProps = {
  searchTerm: string;
  onSearchInput: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onSearchSubmit: (event: React.FormEvent<HTMLFormElement>) => void;
};

const SearchForm = ({
  searchTerm,
  onSearchInput,
  onSearchSubmit,
}: SearchFormProps) => (
  <form onSubmit={onSearchSubmit}>
    <InputWithLabel
      id="search"
      value={searchTerm}
      isFocused
      onInputChange={onSearchInput}
    >
      Search
    </InputWithLabel>
    <button className="button m-3" type="submit" disabled={!searchTerm}>
      Search
    </button>
  </form>
);

export default App;
export { SearchForm, InputWithLabel, List, Item };
