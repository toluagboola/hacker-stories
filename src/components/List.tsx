import React from "react";
import { Check } from "react-feather";
import { Story } from "../App";

type Stories = Array<Story>;

type ListProps = {
	list: Stories;
	onToggleRead: (item: Story) => void;
};

const List = ({ list, onToggleRead }: ListProps) => {
	return (
		<div className="container story-grid">
			{list.map((item) => (
				<Item key={item.objectID} onToggleRead={onToggleRead} item={item} />
			))}
		</div>
	);
};

type ItemProps = {
	item: Story;
	onToggleRead: (item: Story) => void;
};

const Item = ({ item, onToggleRead }: ItemProps) => {
	return (
		<div className="card my-3 has-text-left p-4">
			<p className="mb-2">
				<a className="is-size-6 has-text-weight-medium" href={item.url}>
					{item.title}
				</a>
			</p>

			<p className="pb-1 is-capitalized">
				<strong>Points:</strong> {item.points}
			</p>
			<p className="pb-1 ">
				<strong>Comments:</strong> {item.num_comments}
			</p>
			<button
				type="button"
				className="button is-small mt-2 dismiss"
				onClick={() => onToggleRead(item)}
			>
				<span className="icon mr-2">
					<Check />
				</span>
				{item.isRead ? <span>Mark as unread</span> : <span>Mark as read</span>}
			</button>
		</div>
	);
};

export default List;
export { Item };
