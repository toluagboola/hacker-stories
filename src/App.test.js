// import { render, screen } from "@testing-library/react";
import React from "react";
import axios from "axios";
import renderer from "react-test-renderer";
import App from "./App";
import List, { Item } from "./components/List";
import SearchForm from "./components/SearchForm";

jest.mock("axios");

// Test for Item component
describe("Item", () => {
	const item = {
		title: "React",
		url: "https://reactjs.org/",
		author: "Jordan Walke",
		num_comments: 3,
		points: 4,
		objectID: 0,
	};

	const handleRemoveItem = jest.fn();
	let component;
	beforeEach(() => {
		component = renderer.create(
			<Item item={item} onRemoveItem={handleRemoveItem} />
		);
	});

	it("renders all properties", () => {
		const component = renderer.create(<Item item={item} />);
		expect(component.root.findByType("a").props.href).toEqual(
			"https://reactjs.org/"
		);
	});

	it("calls onRemoveItem on button click", () => {
		component.root.findByType("button").props.onClick();
		expect(handleRemoveItem).toHaveBeenCalledTimes(1);
		expect(handleRemoveItem).toHaveBeenCalledWith(item);
		expect(component.root.findAllByType(Item).length).toEqual(1);
	});

	test("renders snapshot", () => {
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});
});

// Test for List component
describe("List", () => {
	const list = [
		{
			title: "React",
			url: "https://reactjs.org/",
			author: "Jordan Walke",
			num_comments: 3,
			points: 4,
			objectID: 0,
		},
		{
			title: "Redux",
			url: "https://redux.js.org/",
			author: "Dan Abramov, Andrew Clark",
			num_comments: 2,
			points: 5,
			objectID: 1,
		},
	];

	it("renders two times", () => {
		const component = renderer.create(<List list={list} />);
		expect(component.root.findAllByType(Item).length).toEqual(2);
	});
});

// Test for Form component
describe("SearchForm", () => {
	const searchFormProps = {
		searchTerm: "React",
		onSearchInput: jest.fn(),
		onSearchSubmit: jest.fn(),
	};
	let component;

	beforeEach(() => {
		component = renderer.create(<SearchForm {...searchFormProps} />);
	});

	it("renders the input field with its value", () => {
		const value = component.root.findByType("input").props.value;
		expect(value).toEqual("React");
	});

	it("changes the input field", () => {
		const pseudoEvent = { target: "Redux" };
		component.root.findByType("input").props.onChange(pseudoEvent);

		expect(searchFormProps.onSearchInput).toHaveBeenCalledTimes(1);
		expect(searchFormProps.onSearchInput).toHaveBeenCalledWith(pseudoEvent);
	});

	it("submits the form", () => {
		const pseudoEvent = {};
		component.root.findByType("form").props.onSubmit(pseudoEvent);

		expect(searchFormProps.onSearchSubmit).toHaveBeenCalledTimes(1);
		expect(searchFormProps.onSearchSubmit).toHaveBeenCalledWith(pseudoEvent);
	});

	it("disables the button and prevents submit", () => {
		component.update(<SearchForm {...searchFormProps} searchTerm="" />);

		expect(component.root.findByType("button").props.disabled).toBeTruthy();
	});
});

// Test for App component
describe("App", () => {
	it("succeeds fetching data with a list", async () => {
		const list = [
			{
				title: "React",
				url: "https://reactjs.org/",
				author: "Jordan Walke",
				num_comments: 3,
				points: 4,
				objectID: 0,
			},
			{
				title: "Redux",
				url: "https://redux.js.org/",
				author: "Dan Abramov, Andrew Clark",
				num_comments: 2,
				points: 5,
				objectID: 1,
			},
		];

		const promise = Promise.resolve({
			data: {
				hits: list,
			},
		});

		axios.get.mockImplementationOnce(() => promise);
		let component;

		await renderer.act(async () => {
			component = renderer.create(<App />);
		});

		expect(component.root.findByType(List).props.list).toEqual(list);
	});

	it("fails fetching data with a list", async () => {
		const promise = Promise.reject();

		axios.get.mockImplementationOnce(() => promise);

		let component;

		await renderer.act(async () => {
			component = renderer.create(<App />);
		});

		expect(component.root.findByType("p").props.children).toEqual(
			"Something went wrong..."
		);
	});
});

// expect(
// 	component.root.findAllByProps({ children: "Jordan Walke" }).length
// ).toEqual(1);

// test("renders learn react link", () => {
// 	render(<App />);
// 	const linkElement = screen.getByText(/learn react/i);
// 	expect(linkElement).toBeInTheDocument();
// });
