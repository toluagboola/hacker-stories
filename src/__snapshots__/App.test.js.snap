// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item renders snapshot 1`] = `
<div
  className="py-3 has-text-left"
>
  <p>
    <a
      className="is-size-5 has-text-weight-medium"
      href="https://reactjs.org/"
    >
      React
    </a>
  </p>
  <p
    className="pb-1 is-capitalized"
  >
    <strong>
      Author:
    </strong>
     
    Jordan Walke
  </p>
  <p
    className="pb-1 is-capitalized"
  >
    <strong>
      Comments:
    </strong>
     
    3
  </p>
  <p
    className="pb-1 is-capitalized"
  >
    <strong>
      Points:
    </strong>
     
    4
  </p>
  <button
    className="button is-small mt-2 dismiss"
    onClick={[Function]}
    type="button"
  >
    <span
      className="icon"
    >
      <svg
        fill="none"
        height={24}
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        viewBox="0 0 24 24"
        width={24}
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="20 6 9 17 4 12"
        />
      </svg>
    </span>
    <span>
      Check
    </span>
  </button>
</div>
`;
